### PR TITLE
Ensure `RateLimiter` utilizes the token-bucket algorithm

### PR DESCRIPTION
**Pull request by IMax153**
## Type

- [ ] Refactor
- [ ] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Related

- Related Issue #
- Closes #

### CODE REVIEW

File: packages/effect/src/RateLimiter.ts [hunk 1]
```diff
@@ -13,9 +15,11 @@ import * as internal from "./internal/rateLimiter.js"
 import type { Scope } from "./Scope.js"
 
 /**
- * Limits the number of calls to a resource to a maximum amount in some interval using the token bucket algorithm.
+ * Limits the number of calls to a resource to a maximum amount in some interval
+ * using the token-bucket algorithm.
```
Comment (tim-smart) [inline packages/effect/src/RateLimiter.ts:19]: This description is out of date now.

File: packages/effect/test/RateLimiter.test.ts [hunk 2]
```diff
@@ -1,223 +1,271 @@
...
-import * as RateLimiter from "effect/RateLimiter"
 import { assert, describe } from "vitest"
 
-describe("RateLimiterSpec", () => {
-  it.effect("execute up to max calls immediately", () => {
+describe("RateLimiter", () => {
+  describe("fixed-window", () => {
+    RateLimiterTestSuite("fixed-window")
+
+    it.effect("uses the token-bucket algorithm for token replenishment", () =>
```
Comment (tim-smart) [inline packages/effect/test/RateLimiter.test.ts:9]: Should move this to the token bucket block.

File: packages/effect/test/RateLimiter.test.ts [hunk 3]
```diff
@@ -1,223 +1,271 @@
...
+        // Use up all of the available tokens
+        yield* _(Effect.forEach(ReadonlyArray.range(1, 10), () => limit(Effect.unit)))
+
+        // Make an additional request when there are no tokens available
+        yield* _(
+          limit(Effect.unit),
+          Effect.zipRight(Deferred.succeed(deferred, void 0)),
+          Effect.fork
+        )
+        assert.isFalse(yield* _(Deferred.isDone(deferred)))
```
Comment (tim-smart) [inline packages/effect/test/RateLimiter.test.ts:30]: You will technically need a yieldNow before this, as forks don't execute immediately.

**Review by tim-smart — CHANGES_REQUESTED** [review]

File: packages/effect/src/RateLimiter.ts [hunk 4]
```diff
@@ -27,11 +31,49 @@ export interface RateLimiter {
...
+    /**
+     * The interval to utilize for rate-limiting requests. The semantics of the
+     * specified `interval` vary depending on the chosen `algorithm`:
+     *
+     * `token-bucket`: The maximum number of requests will be spread out over
+     * the provided interval if no tokens are available.
+     *
+     * For example, for a `RateLimiter` using the `token-bucket` algorithm with
+     * a `limit` of `10` and an `interval` of `1 seconds`, `1` request can be
+     * made every `100 millis`.
```
Comment (hsubra89) [inline packages/effect/src/RateLimiter.ts:55]:
```suggestion
     * For example, for a `RateLimiter` using the `token-bucket` algorithm with
     * a `limit` of `10` and an `interval` of `1 seconds`, `1` request can be
     * made every `100 millis` when the token bucket is empty.
```

File: packages/effect/src/internal/rateLimiter.ts [hunk 5]
```diff
@@ -1,14 +1,53 @@
...
+    const millisPerToken = Math.ceil(Duration.toMillis(window) / limit)
+    const queue = yield* _(Queue.bounded<void>(limit - 1))
+    yield* _(
+      queue.take,
+      Effect.zipRight(Effect.sleep(millisPerToken)),
+      Effect.forever,
+      Effect.forkScoped,
+      Effect.interruptible
+    )
+    return (effect) => Effect.zipRight(queue.offer(void 0), effect)
```
Comment (hsubra89) [inline packages/effect/src/internal/rateLimiter.ts:44]: It took me a while to understand this, but this is certainly very elegant :)

File: packages/effect/test/RateLimiter.test.ts (hunk 3 above)
Comment (IMax153) [reply packages/effect/test/RateLimiter.test.ts:34]: Good call - fixed 👍

File: packages/effect/test/RateLimiter.test.ts (hunk 2 above)
Comment (IMax153) [reply packages/effect/test/RateLimiter.test.ts:13]: That was dumb - thanks 👍

File: packages/effect/src/RateLimiter.ts (hunk 1 above)
Comment (IMax153) [reply packages/effect/src/RateLimiter.ts]: Good call - fixed 👍

**Comment by IMax153** [comment]
@tim-smart @mikearnaldi - since we're now updating the public signature of `RateLimiter.make`, should this be a minor?

**Review by tim-smart — APPROVED** [review]

**Pull request merged**